From fcbbee5902817d1a9a24da75381dbb20b17de980 Mon Sep 17 00:00:00 2001
From: Girish S G <girishsg@ti.com>
Date: Tue, 1 Mar 2011 12:52:50 -0600
Subject: OMAP4:PM:Regulator suspend state behavior

This introduces support for setting the required states
in suspend. Right now disabling it in suspend.
Respective driver would be taking care of enabling it when
needed.
REVISIT might be needed as required.

Change-Id: Ia4b6f678949a044279b816c46b78a41b3496aabc
Signed-off-by: Girish S G <girishsg@ti.com>
---
 arch/arm/mach-omap2/board-4430sdp.c |   40 +++++++++++++++++++++++++++++++++++
 drivers/regulator/twl-regulator.c   |   18 +++++++++++++++
 2 files changed, 58 insertions(+), 0 deletions(-)

Index: kernel/arch/arm/mach-omap2/board-4430sdp.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/board-4430sdp.c	2011-04-01 18:30:35.000000000 +0900
+++ kernel/arch/arm/mach-omap2/board-4430sdp.c	2011-04-02 15:56:19.000000000 +0900
@@ -869,6 +869,10 @@
 		.valid_ops_mask	 = REGULATOR_CHANGE_VOLTAGE
 					| REGULATOR_CHANGE_MODE
 					| REGULATOR_CHANGE_STATUS,
+		.state_mem = {
+			.enabled	= false,
+			.disabled	= true,
+		},
 	},
 };
 
@@ -895,6 +899,10 @@
 		.valid_ops_mask	 = REGULATOR_CHANGE_VOLTAGE
 					| REGULATOR_CHANGE_MODE
 					| REGULATOR_CHANGE_STATUS,
+		.state_mem = {
+			.enabled	= false,
+			.disabled	= true,
+		},
 	},
 	.num_consumer_supplies = 1,
 	.consumer_supplies = sdp4430_cam2_supply,
@@ -911,6 +919,10 @@
 		.valid_ops_mask	 = REGULATOR_CHANGE_VOLTAGE
 					| REGULATOR_CHANGE_MODE
 					| REGULATOR_CHANGE_STATUS,
+		.state_mem = {
+			.enabled	= false,
+			.disabled	= true,
+		},
 	},
 	.num_consumer_supplies  = 1,
 	.consumer_supplies      = sdp4430_vmmc_supply,
@@ -926,6 +938,10 @@
 		.valid_ops_mask	 = REGULATOR_CHANGE_VOLTAGE
 					| REGULATOR_CHANGE_MODE
 					| REGULATOR_CHANGE_STATUS,
+		.state_mem = {
+			.enabled	= false,
+			.disabled	= true,
+		},
 	},
 };
 
@@ -939,6 +955,10 @@
 		.valid_ops_mask	 = REGULATOR_CHANGE_VOLTAGE
 					| REGULATOR_CHANGE_MODE
 					| REGULATOR_CHANGE_STATUS,
+		.state_mem = {
+			.enabled	= false,
+			.disabled	= true,
+		},
 	},
 };
 
@@ -951,6 +971,10 @@
 					| REGULATOR_MODE_STANDBY,
 		.valid_ops_mask	 = REGULATOR_CHANGE_MODE
 					| REGULATOR_CHANGE_STATUS,
+		.state_mem = {
+			.enabled	= false,
+			.disabled	= true,
+		},
 	},
 };
 
@@ -963,6 +987,10 @@
 					| REGULATOR_MODE_STANDBY,
 		.valid_ops_mask	 = REGULATOR_CHANGE_MODE
 					| REGULATOR_CHANGE_STATUS,
+		.state_mem = {
+			.enabled	= false,
+			.disabled	= true,
+		},
 	},
 };
 
@@ -975,6 +1003,10 @@
 					| REGULATOR_MODE_STANDBY,
 		.valid_ops_mask	 = REGULATOR_CHANGE_MODE
 					| REGULATOR_CHANGE_STATUS,
+		.state_mem = {
+			.enabled	= false,
+			.disabled	= true,
+		},
 	},
 };
 
@@ -987,6 +1019,10 @@
 					| REGULATOR_MODE_STANDBY,
 		.valid_ops_mask	 =	REGULATOR_CHANGE_MODE
 					| REGULATOR_CHANGE_STATUS,
+		.state_mem = {
+			.enabled	= false,
+			.disabled	= true,
+		},
 	},
 };
 
Index: kernel/drivers/regulator/twl-regulator.c
===================================================================
--- kernel.orig/drivers/regulator/twl-regulator.c	2011-04-01 18:30:43.000000000 +0900
+++ kernel/drivers/regulator/twl-regulator.c	2011-04-02 15:50:15.000000000 +0900
@@ -383,6 +383,20 @@
 	return LDO_MV(info->table[vsel]) * 1000;
 }
 
+static int twlreg_suspend_enable(struct regulator_dev *rdev)
+{
+	return twlreg_enable(rdev);
+}
+
+static int twlreg_suspend_disable(struct regulator_dev *rdev)
+{
+	/*
+	 * In suspend disable, LDO's sleep state can be configured. Right
+	 * now disabling ldo's at suspend.
+	 */
+	return twlreg_disable(rdev);
+}
+
 static struct regulator_ops twl4030ldo_ops = {
 	.list_voltage	= twl4030ldo_list_voltage,
 
@@ -452,6 +466,8 @@
 	.set_mode	= twlreg_set_mode,
 
 	.get_status	= twlreg_get_status,
+	.set_suspend_enable = twlreg_suspend_enable,
+	.set_suspend_disable = twlreg_suspend_disable,
 };
 
 /*----------------------------------------------------------------------*/
@@ -485,6 +501,8 @@
 	.set_mode	= twlreg_set_mode,
 
 	.get_status	= twlreg_get_status,
+	.set_suspend_enable = twlreg_suspend_enable,
+	.set_suspend_disable = twlreg_suspend_disable,
 };
 
 /*----------------------------------------------------------------------*/
